### PR TITLE
Upgrade script

### PR DESCRIPTION
**Pull request by Psychi1**
Hi,

this script helps to upgrade snappymail according to the [general instructions](https://github.com/the-djmaze/snappymail/wiki/Installation-instructions#general-instructions) from your wiki. 

Tested with Ubuntu 20.04 LTS.

Bye

### CODE REVIEW

**Review by the-djmaze — CHANGES_REQUESTED** [review]
This script is a good start but is too much related to your setup.
For example i use /home/snappymail and run it as isolated chroot user.

The script also scrapes to find the tar.gz which should get a better system.
Like i provide a snappymail.eu/releases/latest.tar.gz or something.

I could help improve the script, but currently i can't approve.

**Comment by Psychi1** [comment]
Hello,

thanks for your reply and your suggestions.

> This script is a good start but is too much related to your setup. For example i use /home/snappymail and run it as isolated chroot user.

I followed the install instructions from your [wiki](https://github.com/the-djmaze/snappymail/wiki/Installation-instructions#general-instructions), which refers to `/var/www/` as document root. Of course it's only a dictionary like `/home/snappymail`. 

I don't have any experience with chroot. If it's not working like bash, I have to give up.

> The script also scrapes to find the tar.gz which should get a better system. Like i provide a snappymail.eu/releases/latest.tar.gz or something.

This is a really good point. I don't like the scrapping. However you are accountable for the name of the release file. When a `latest.tar.gz` is created, it will simplify the script.

> I could help improve the script, but currently i can't approve this request.

Help is welcome.

**Comment by jas8522** [comment]
I'm not entirely sure how to suggest changes to another user's pull request, so attached is a version using the new latest.tar.gz URL and which removes the environment specific code
[upgrade_snappymail.txt](https://github.com/the-djmaze/snappymail/files/8328467/upgrade_snappymail.txt)
.

**Comment by the-djmaze** [comment]
@jas8522 thanks.

However, I noticed a problem that also resides in @Psychi1 code.

```
OLD_VERSION=`cat data/VERSION`
```

You may configure `data/VERSION` to a different location and then the file can't be found.
The only reliable source for current version is `index.php` line:
```
define('APP_VERSION', '0.0.0');
```

**Comment by jas8522** [comment]
Updated to:
```
OLD_VERSION=`grep "define('APP_VERSION" index.php | awk -F\' '{print $4}'`
```

[upgrade_snappymail.txt](https://github.com/the-djmaze/snappymail/files/8332948/upgrade_snappymail.txt)

**Comment by Psychi1** [comment]
Hello,

thanks to @jas8522 for your ideas and to @the-djmaze for the _lastest_ link. I changed to code according to @jas8522 suggestions. However I added to colors again ... I'm a color guy. So the user does not miss the importent information.

**Comment by jas8522** [comment]
@Psychi1 No prob! I used it for my own updating today and it worked great :)

No complaints about the colours here; I just tend to prefer code readability when possible. Then again half of shell scripting is hard to parse, so that's pretty much a non-issue ;)

**Comment by jas8522** [comment]
Oh something that crossed my mind: if snappymail is installed in a subdirectory of the web root, the backups this upgrade tool creates will be web accessible and constitute a security risk. I don't *think* we can fix that programmatically, but perhaps it should output a warning after the echo showing it's backing up indicating to ensure the parent directory is not web accessible. Thoughts?

**Comment by Psychi1** [comment]
Hi, 

the backups are store one level above the root. So it depends on the webserver configuration to deny access. 

We can introduce an ENV like SNAPPY_BACKUP_FOLDER and save - if set - the backup there or - if not set - backup as usual with a warning message.

**Comment by eibex** [comment]
> Alternatively we could exclude the data folder, but then it wouldn't be a complete file backup, and I suspect people would expect it to be complete.

You could argue that since the data folder is never touched during the upgrade process it would be okay not to back it up. As a rule of thumb you shouldn't keep it under the webroot, but in an external directory.

But I do agree that backups shouldn't be at risk of being publicly accessible. For this reason, I turned the backing up string of this script off for a while now (since my data directory wouldn't be affected in case of errors).

**Comment by jas8522** [comment]
How about this: we change it to backup to /tmp instead, then at the end of the script indicate that if anything is amiss, the backup can temporarily be found at /tmp/FILENAME. That way it's sure to be out of the web root yet a backup is still created for near-immediate recoveries. I think for longer-term backup solutions, the user should either be manually backing up anyway, or already have an automated backup solution in place.

**Comment by the-djmaze** [comment]
> How about this: we change it to backup to /tmp instead

/tmp is not /tmp on some systems.
There are cases where each PHP-FPM has its own hidden /tmp (like in cPanel).
Most (shared hosting) users will have no idea where the backup will be (or even accessible).

**Comment by jas8522** [comment]
Does this matter much anymore now that there's a built in updater from the admin?

**Comment by the-djmaze** [comment]
> Does this matter much anymore now that there's a built in updater from the admin?

It does matter.
Not everyone has ?admin enabled and never checks updates there.

**Comment by jas8522** [comment]
For consistency, could/should a CLI script like this trigger the PHP run by the admin UI updater?

**Comment by master3395** [comment]
> > This is a really good point. I don't like the scrapping. However, you are accountable for the name of the release file. When a `latest.tar.gz` is created, it will simplify the script.
> 
> Done. Locations are https://snappymail.eu/repository/latest.tar.gz

Could we perhaps also have one with just zip?
https://snappymail.eu/repository/latest.zip

**Comment by the-djmaze** [comment]
> Could we perhaps also have one with just zip?

Preferably not. Problem with zip is the lack of metadata (like file permissions).

**Comment by Zero3K** [comment]
How about having the script download a command line tool that extracts tar.gz and extracts it using the tool?

**Comment by the-djmaze** [comment]
@Zero3K like here:

https://github.com/the-djmaze/snappymail/blob/4def341b66d91e213848a2e3037f7d391f085abf/snappymail/v/0.0.0/app/libraries/snappymail/upgrade.php#L197-L206

**Comment by ancmto** [comment]
In the settings could have the box to check the option "Reply to all" as default.

In the corporate environment, this option is used a lot, and having this option set as default would be useful.

**Comment by the-djmaze** [comment]
> What needs changing before merge?

Not much, so i've merged it and moved it /bin/upgrade.sh

**Comment by DeeFuse** [comment]
You could get the latest .tar.gz asset from GitHub by using this snippet:
`LATEST_URL=$(curl -ks https://api.github.com/repos/the-djmaze/snappymail/releases/latest | grep "browser_download_url.*.tar.gz\"" | cut -d '"' -f 4)`
